pkg/applicationserver/io/packages/alcsync/v1/timesync.go

```diff
@@ -0,0 +1,59 @@
+// Copyright © 2023 The Things Network Foundation, The Things Industries B.V.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package alcsyncv1
+
+import (
+	"context"
+	"time"
+
+	"go.thethings.network/lorawan-stack/v3/pkg/errors"
+	"go.thethings.network/lorawan-stack/v3/pkg/ttnpb"
+)
+
+var errUnknownCommand = errors.DefineNotFound("unknown_command", "unknown command")
+
+// TimeSyncCommand is the command for time synchronization.
+type TimeSyncCommand struct {
+	req              *AppTimeReq
+	receivedAt       time.Time
+	thresholdSeconds uint32
+	fPort            uint32
+}
+
+// Code implements commands.Command.
+func (*TimeSyncCommand) Code() uint8 {
+	return 0x01
+}
+
+// Execute implements commands.Command.
+func (cmd *TimeSyncCommand) Execute(
+	ctx context.Context,
+	onDownlink func(context.Context, *ttnpb.ApplicationDownlink) error,
+) error {
+	difference := cmd.receivedAt.Sub(cmd.req.DeviceTime)
+	timeCorrectionSeconds := int32(difference.Seconds())
+	ans := &AppTimeAns{
+		TimeCorrection: timeCorrectionSeconds,
+		TokenAns:       cmd.req.TokenReq,
+	}
+	frmPayload := ans.Marshal()
+	if cmd.req.AnsRequired || ans.TimeCorrection >= int32(cmd.thresholdSeconds) {
```

Comment:
Shouldn't we compare the absolute value of the time correction with the threshold ? Being in the past is an error too.

Reply:
True.